sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1518,6 +1518,163 @@ case class ArrayContains(left: Expression, right: Expression)
     copy(left = newLeft, right = newRight)
 }
 
+/**
+ * Searches the specified array for the specified object using the binary search algorithm.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(array, value) - Return index (0-based) of the search value, " +
+    "if it is contained in the array; otherwise, (-<insertion point> - 1).",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), 2);
+       1
+      > SELECT _FUNC_(array(null, 1, 2, 3), 2);
+       2
+  """,
+  group = "array_funcs",
+  since = "4.0.0")
+// scalastyle:on line.size.limit
+case class ArrayBinarySearch(array: Expression, value: Expression)
```

Comment:
Shouldn't this be called ArrayIndexOf or something like that? That this is implemented using binary search is not that interesting for an end-user.

Reply:
I think we're just using it internally within Spark (to implement `function plotting` for the `connect` module)